### PR TITLE
[GooglePlayReleaseV4] Add option to upload native debug symbols

### PR DESCRIPTION
**Pull request by DaniilShmelev**
**Task name**: GooglePlayReleaseV4

**Description**: Added an optional input that allows uploading a native debug symbols zip archive along with release.

**Documentation changes required:** Yes

**Added unit tests:** No

**Attached related issue:** https://github.com/microsoft/google-play-vsts-extension/issues/249

**Checklist**:
- [x] Task version was bumped - please check [instruction](https://github.com/microsoft/google-play-vsts-extension/tree/master/docs/taskversionbumping.md) how to do it
- [x] Checked that applied changes work as expected

### CODE REVIEW

File: Tasks/GooglePlayReleaseV4/modules/googleutil.ts [hunk 1]
```diff
@@ -265,6 +265,43 @@ export async function uploadDeobfuscation(
...
+    };
+
+    try {
+        tl.debug('Request Parameters: ' + JSON.stringify(requestParameters));
+        const res = await edits.deobfuscationfiles.upload(requestParameters, { onUploadProgress });
+        tl.debug('returned: ' + JSON.stringify(res));
+        return res.data;
+    } catch (e) {
+        tl.debug(`Failed to upload native deobfuscation file ${mappingFilePath}`);
+        tl.debug(e);
```
Comment (anatolybolshakov) [inline Tasks/GooglePlayReleaseV4/modules/googleutil.ts:300]: Do we really need to print the error twice?

**Review by anatolybolshakov — APPROVED** [review]
LGTM, thanks! Please take a look at comment

File: Tasks/GooglePlayReleaseV4/modules/googleutil.ts (hunk 1 above)
Comment (DaniilShmelev) [reply Tasks/GooglePlayReleaseV4/modules/googleutil.ts:300]: Yeah, there is some redundancy in these error messages. Fixed, thanks!

**Review by denis-tikhomirov — APPROVED** [review]

File: Tasks/GooglePlayReleaseV4/main.ts [hunk 2]
```diff
@@ -69,6 +69,9 @@ async function run(): Promise<void> {
         const uploadMappingFile: boolean = tl.getBoolInput('shouldUploadMappingFile', false) && (action === 'SingleApk' || action === 'SingleBundle');
         const mappingFilePattern: string = tl.getInput('mappingFilePath');
 
+        const uploadNativeMappingFile: boolean = tl.getBoolInput('shouldUploadNativeMappingFile', false) && (action === 'SingleApk' || action === 'SingleBundle');
+        const nativeMappingFilePattern: string = tl.getInput('nativeMappingFilePath');
```
Comment (alexander-smolyakov) [inline Tasks/GooglePlayReleaseV4/main.ts:73]: Could you add information about the new task inputs to the documentation?

File: Tasks/GooglePlayReleaseV4/main.ts [hunk 3]
```diff
@@ -165,6 +168,16 @@ async function run(): Promise<void> {
...
             }
+
+            if (uploadNativeMappingFile) {
+                tl.debug(`Native mapping file pattern: ${nativeMappingFilePattern}`);
+
+                const mappingFilePath = fileHelper.resolveGlobPath(nativeMappingFilePattern);
+                tl.checkPath(mappingFilePath, 'Native mapping file path');
+                console.log(tl.loc('FoundNativeDeobfuscationFile', mappingFilePath));
+                tl.debug(`Uploading ${mappingFilePath} for version code ${versionCodes[0]}`);
+                await googleutil.uploadNativeDeobfuscation(edits, mappingFilePath, packageName, versionCodes[0]);
```
Comment (alexander-smolyakov) [inline Tasks/GooglePlayReleaseV4/main.ts:179]:
```suggestion
                tl.debug(`Native mapping file pattern: ${nativeMappingFilePattern}`);
                
                const mappingFilePath: string = fileHelper.resolveGlobPath(nativeMappingFilePattern);
                tl.checkPath(mappingFilePath, 'Native mapping file path');
                
                console.log(tl.loc('FoundNativeDeobfuscationFile', mappingFilePath));
                tl.debug(`Uploading ${mappingFilePath} for version code ${versionCodes[0]}`);
                await googleutil.uploadNativeDeobfuscation(edits, mappingFilePath, packageName, versionCodes[0]);
```

File: Tasks/GooglePlayReleaseV4/modules/googleutil.ts [hunk 4]
```diff
@@ -259,12 +259,45 @@ export async function uploadDeobfuscation(
...
+        packageName: packageName,
+        apkVersionCode: versionCode,
+        media: {
+            body: fs.createReadStream(mappingFilePath),
+            mimeType: ''
+        }
+    };
+
+    try {
+        tl.debug('Request Parameters: ' + JSON.stringify(requestParameters));
```
Comment (alexander-smolyakov) [inline Tasks/GooglePlayReleaseV4/modules/googleutil.ts:292]: Could you double-check whether it is really necessary to output all request parameters to the logs? Is there any possibility that cold request parameters contain sensitive data or PII data? 

Also, nit:
```suggestion
        tl.debug(`Request Parameters: ${JSON.stringify(requestParameters)}`);
```

File: Tasks/GooglePlayReleaseV4/modules/googleutil.ts [hunk 5]
```diff
@@ -259,12 +259,45 @@ export async function uploadDeobfuscation(
...
+        const res = await edits.deobfuscationfiles.upload(requestParameters, { onUploadProgress });
```
Comment (alexander-smolyakov) [inline Tasks/GooglePlayReleaseV4/modules/googleutil.ts:293]: Could you please set type for `res` variable?

File: Tasks/GooglePlayReleaseV4/modules/googleutil.ts [hunk 6]
```diff
@@ -259,12 +259,45 @@ export async function uploadDeobfuscation(
...
+        tl.debug('returned: ' + JSON.stringify(res));
```
Comment (alexander-smolyakov) [inline Tasks/GooglePlayReleaseV4/modules/googleutil.ts:294]: nit:
```suggestion
        tl.debug(`Response: ${JSON.stringify(res)}`);
```

**Review by alexander-smolyakov — COMMENTED** [review]
Could you please take a look at the comments?

File: Tasks/GooglePlayReleaseV4/modules/googleutil.ts (hunk 4 above)
Comment (DaniilShmelev) [reply Tasks/GooglePlayReleaseV4/modules/googleutil.ts:292]: Request parameters don't contain any secret info, but may be useful for debug purposes.
The only customer's info logged here is package name, version code and debug symbols path. All of these are passed to the task as non-secret inputs.

File: Tasks/GooglePlayReleaseV4/modules/googleutil.ts (hunk 5 above)
Comment (DaniilShmelev) [reply Tasks/GooglePlayReleaseV4/modules/googleutil.ts:293]: The response type here is long and doesn't add too much to code readability: `googleapis.Common.GaxiosResponse<googleapis.androidpublisher_v3.Schema$DeobfuscationFilesUploadResponse>`.
In all other functions of this module, `res` variable doesn't have an explicit type. Consistency is another reason to keep this type implicit.

File: Tasks/GooglePlayReleaseV4/main.ts (hunk 3 above)
Comment (DaniilShmelev) [reply Tasks/GooglePlayReleaseV4/main.ts]: Fixed in https://github.com/microsoft/google-play-vsts-extension/pull/330/commits/452d8689a05995612c8ce92be99826dca89e2f7e

File: Tasks/GooglePlayReleaseV4/main.ts (hunk 2 above)
Comment (DaniilShmelev) [reply Tasks/GooglePlayReleaseV4/main.ts:73]: Added in https://github.com/microsoft/google-play-vsts-extension/pull/330/commits/5e97e0a2151789e78228cebac93e79914626b242

**Review by alexander-smolyakov — APPROVED** [review]
LGTM, thanks!

File: Tasks/GooglePlayReleaseV4/Strings/resources.resjson/en-US/resources.resjson [hunk 7]
```diff
@@ -43,6 +43,10 @@
   "loc.input.help.shouldUploadMappingFile": "Select this option to attach your proguard mapping.txt file to your aab/apk.",
   "loc.input.label.mappingFilePath": "Deobfuscation path",
   "loc.input.help.mappingFilePath": "The path to the proguard mapping.txt file to upload. Glob patterns are supported.",
+  "loc.input.label.shouldUploadNativeMappingFile": "Upload native debug symbols",
+  "loc.input.help.shouldUploadNativeMappingFile": "Select this option to attach your native mapping file to your aab/apk.",
+  "loc.input.label.nativeMappingFilePath": "Native debug symbols file path",
```
Comment (max-zaytsev) [inline Tasks/GooglePlayReleaseV4/Strings/resources.resjson/en-US/resources.resjson:48]: I think the word "mapping" can confuse someone here.
Can we consider renaming these inputs and variables like "nativeDebugSymbolsFile..."?

Comment (DaniilShmelev) [reply Tasks/GooglePlayReleaseV4/Strings/resources.resjson/en-US/resources.resjson]: Fixed, thanks

**Comment by max-zaytsev** [comment]
Thanks for doing this!

**Review by max-zaytsev — APPROVED** [review]

**Comment by DaniilShmelev** [comment]
@mmrazik could you please review and merge this?

**Review by mmrazik — APPROVED** [review]

**Pull request merged**